Add option to OpInfo to skip gradgrad check and empty cdist OpInfo

albanD opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #56606 Fix cdist backward for empty inputs
* #56605 Fix broadcasting of cdist backward
* #56604 Add more test cases for cdist OpInfo and TODOs
* **#56603 Add option to OpInfo to skip gradgrad check and empty cdist OpInfo**

Differential Revision: [D27939204](https://our.internmc.facebook.com/intern/diff/D27939204)

albanD committed:
Add option to OpInfo to skip gradgrad check and empty cdist OpInfo

[ghstack-poisoned]

albanD committed:
Update on "Add option to OpInfo to skip gradgrad check and empty cdist OpInfo"




[ghstack-poisoned]

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -234,6 +235,7 @@ def __init__(self,
             self.supports_complex_autograd = supports_autograd
 
         self.gradcheck_wrapper = gradcheck_wrapper
+        self.check_gradgrad = check_gradgrad
```
Would "supports_gradgrad" be more accurate?

If so, what about modeling the default value as None and inheriting from `supports_autograd` by default? Currently if someone set `supports_autograd=False`, for example, the `check_gradgrad` would still be `True`.

albanD replied:
I wasn't sure how to name it. I used this to match the existing `check_batch_grad` and `check_batch_gradgrad` naming convention.

All 3 of these args are only checking if `supports_autograd=True` and ignored otherwise.

mruberry replied:
OK; let's just go with "supports_gradgrad", which is a statement about the operator (whereas check_batch_grad is a testing directive) and just extend the comment to note that the value is ignored if supports_autograd is false

cool?

albanD replied:
Sounds good!

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -3293,6 +3301,15 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
+           supports_out=False,
+           check_gradgrad=False,
+           skips=(
+               # jit consistency always checks gradgrad
```
Can we add the check to the test along with a `# TODO: ...` comment explaining that it'd be nice if the test could avoid gradgrad checks if the op doesn't support gradgrad?

albanD replied:
I wanted to update it, but it happens here: https://github.com/pytorch/pytorch/blob/5cab3b9cf628ba44b599fa7da4c06404986dd656/torch/testing/_internal/common_jit.py#L40
Which is jit specific code. And my quick attempt to modify it lead to more things blowing up in other tests.

But if you think that we should add it there, happy to do it. Just let me know where it would be most appropriate.

mruberry replied:
I think what we can do for is update this logic:

https://github.com/pytorch/pytorch/blob/9be2cabc4575f86e53c2b66ea277084f5c003c0a/test/test_ops.py#L346

so that requires grad is only true if both grad and grad grad are supported, and add the comment there explaining why this is imperfect (because we could still test grad and not grad grad if the jit helper was updated to understand that nuance)

albanD replied:
That what I tried, this is what it leads to:
```
torch._C._jit_check_alias_annotation(CU.the_method.graph, tuple(tensors), aten_name)
RuntimeError: aliasOp != torch::jit::getOperatorAliasMap().end()INTERNAL ASSERT FAILED at "../torch/csrc/jit/passes/utils/check_alias_annotation.cpp":155, please report a bug to PyTorch.
```

That's why I just disabled it :D

mruberry replied:
Huh, that's bizarre that setting _requires_grad to False here results in an aliasing error. 

In that case, would you just skip in the test:

```
# TODO: fix this
if _requires_grad and not op.supports_gradgrad:
  self.skipTest("skipped! Test does not handle ops that don't support gradgrad properly!")
```

?

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -3293,6 +3301,15 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
                    dtypesIfCUDA=None,
                    dtypesIfROCM=None,
                    supports_autograd=False),
+    OpInfo('cdist',
```
Should this get a `# TODO: ...` to subsume the cdist tests in test_autograd.py?

https://github.com/pytorch/pytorch/blob/5017c5fcad5889acba1623c634d0aaeadb676aa0/test/test_autograd.py#L7312

albanD replied:
No, unfortunately, all the tests in test_autograd check different things. Mostly that we get the right output/gradient for boundary values and for very large inputs.
The ones from test_torch that can be moved are moved in the next PR in the stack.

mruberry reviewed (commented):
Hey @albanD! This is cool! I added a couple questions; looking forward to hearing your thoughts

albanD committed:
Update on "Add option to OpInfo to skip gradgrad check and empty cdist OpInfo"


Differential Revision: [D27939204](https://our.internmc.facebook.com/intern/diff/D27939204)

[ghstack-poisoned]

mruberry reviewed (approved):
Cool!

I made one last suggestion to try and move the skip in test_variant_consistency_jit itself for ops where supports_gradgrad is False. If that doesn't work, however, this PR's current approach of just skipping the test is OK, too.

albanD committed:
Update on "Add option to OpInfo to skip gradgrad check and empty cdist OpInfo"


Differential Revision: [D27939204](https://our.internmc.facebook.com/intern/diff/D27939204)

[ghstack-poisoned]
